partial/pandas/_libs/interval.pyi

```diff
@@ -29,43 +40,59 @@ class IntervalMixin:
     @property
     def is_empty(self: IntervalMixinProtocol) -> bool: ...
 
-class Interval(IntervalMixin, Generic[Orderable]):
+class Interval(IntervalMixin, Generic[OrderableT]):
     @overload
-    def left(self: Interval[OrderableScalar]) -> OrderableScalar: ...
+    def left(self: Interval[OrderableScalarT]) -> OrderableScalar: ...
     @overload
-    def left(self: Interval[OrderableTimes]) -> OrderableTimes: ...
+    def left(self: Interval[OrderableTimesT]) -> OrderableTimes: ...
     @property
     @overload
-    def left(self: Interval[Orderable]) -> Orderable: ...
+    def left(self: Interval[OrderableT]) -> Orderable: ...
     @overload
-    def right(self: Interval[OrderableScalar]) -> OrderableScalar: ...
+    def right(self: Interval[OrderableScalarT]) -> OrderableScalar: ...
     @overload
-    def right(self: Interval[OrderableTimes]) -> OrderableTimes: ...
+    def right(self: Interval[OrderableTimesT]) -> OrderableTimes: ...
     @property
     @overload
-    def right(self: Interval[Orderable]) -> Orderable: ...
+    def right(self: Interval[OrderableT]) -> Orderable: ...
     @property
     def closed(self) -> str: ...
-    def __init__(
-        self,
-        left: Orderable,
-        right: Orderable,
+    @overload
+    def __new__(
+        cls,
+        left: Timestamp,
+        right: Timestamp,
         closed: Union[str, Literal["left", "right", "both", "neither"]] = ...,
-    ) -> None: ...
+    ) -> Interval[Timestamp]: ...
     @overload
-    def length(self: Interval[OrderableScalar]) -> float: ...
+    def __new__(
+        cls,
+        left: Timedelta,
+        right: Timedelta,
+        closed: Union[str, Literal["left", "right", "both", "neither"]] = ...,
+    ) -> Interval[Timedelta]: ...
     @overload
-    def length(self: Interval[OrderableTimes]) -> Timedelta: ...
-    @property
+    def __new__(
+        cls,
+        left: int,
+        right: int,
+        closed: Union[str, Literal["left", "right", "both", "neither"]] = ...,
+    ) -> Interval[int]: ...
     @overload
-    def length(self: Interval[Orderable]) -> Orderable: ...
+    def __new__(
```

Comment:
and the same for `__new__`

```py
    def __new__(
        cls,
        left: OrderableT,
        right: OrderableT,
        closed: Literal["left", "right", "both", "neither"] = ...,
    ) -> Interval[OrderableT]: ...
```

Reply:
All of them are needed.  That's how you end up with the specific matching of input types to the proper `Interval[]` subtype. I tried removing them and then `length` has the wrong type.